### ISSUE TITLE
`convert` for `GAP.GapObj`

### ISSUE THREAD

**Issue by thofma** [oscar-system/Oscar.jl#681]
Is there a reason to define `Base.convert(GAP.GapObj, ...)`? Since `GAP.GapObj(x)` is not necessarily of type `GAP.GapObj` this is breaking julia left and right if julia really makes use of the `convert` functions.

```julia
julia> x = Oscar.GAP.GapObj[]
GAP_jll.MPtr[]

julia> push!(x, fmpq(1)//2)
1-element Vector{GAP.GapObj}:
 GAP: 1/2

julia> push!(x, fmpq(1))
ERROR: TypeError: in typeassert, expected GAP.GapObj, got a value of type Int64
Stacktrace:
 [1] setindex!(A::Vector{GAP.GapObj}, x::Int64, i1::Int64)
   @ Base ./array.jl:839
 [2] push!(a::Vector{GAP.GapObj}, item::fmpq)
   @ Base ./array.jl:930
 [3] top-level scope
   @ REPL[18]:1
```

Since the only application of `convert` is to have the automatic conversion for the `setindex` calls and they are obviously broken, I don't see why we have those `convert` methods at all.

@fingolfin @ThomasBreuer 

P.S.: I would suggest to remove them all. `convert` is dangerous and `GAPObj` is lowlevel enough to not rely on automatic conversions.

**Comment by ThomasBreuer** [comment]
Concerning the fact that `GAP.GapObj(x)` currently does not necessarily return an object of type `GAP.GapObj(x)`, this is something which will be changed, see oscar-system/GAP.jl/issues/689.
Concerning `convert` methods, we had provided them in GAP.jl in addition to constructor methods, referring to https://discourse.julialang.org/t/4159/24 and https://discourse.julialang.org/t/4159/35.
Should we really remove all these methods?

**Comment by thofma** [comment]
Good that it will be made consistent. It was not clear to me that a decision has been made for https://github.com/oscar-system/GAP.jl/issues/689.

Independent of this, I don't understand why one would want to implement `convert` in this case. Users should not use `convert` but the type constructors. Is the implicit call to `convert` (which is the only interesting consequence of defining `convert`) used a lot? In my experience the presence of `convert` methods makes it easy to write buggy code (see the tuple vs. vector julia exercise from last week).

**Comment by fingolfin** [comment]
Personally I never use `convert(T, x)` in this context, I always use `T(x)`.

The main "advantages" of `convert` seem to be:
- saying `convert(T,x)` may convey what's happening better to people, esp. those coming from outside Julia (but they need to learn about `T(x)` at some point anyway)
- it is used for various implicit conversions -- whether that's a reason for using it or against using it depends on use cases, though... See also https://github.com/oscar-system/Oscar.jl/pull/680/files#r708175881

I wonder if @ThomasBreuer or @FriedrichRober have any use cases for the implicit conversion, or anything else to add?

**Comment by FriedrichRober** [comment]
I can't think of any use case for implicit conversion atm. I don't have a strong opinion on this, but it seems to be probably easier for code maintenance to drop all `convert(T, x)` methods for `GapObj`.

**Comment by ThomasBreuer** [comment]
Perhaps we should look at the history of the development:

First we had only `gap_to_julia`, which returns an object of the given type or of a default type, and converts recursively or just the outer level, depending on the `recursive` argument.
Then we introduced `convert` methods, where the Julia type cannot be omitted, and again with the optional `recursive` argument.
And we introduced the constructor methods for the Julia types in question, also with optional `recursive`.

Concerning the latter two, I am also using the constructors not `convert`. My understanding was that the more programmatic `convert` would yield some general benefits in Julia, like the mentioned implicit conversions. If `convert` is deprecated, we can remove the methods in question from GAP.jl.

In order to make explicit where conversions from GAP to Julia happen, I would use `gap_to_julia`. In a sense, the question is what one wants to express: If the GAP objects are regarded mainly as Julia objects, applying constructors is appropriate, and if GAP objects are regarded as "foreign" objects (for which expensive conversions are needed before they can be used), this can be expressed by calling `gap_to_julia`.
